_posts/2022-9-5-Diplomat @ Google.md

```diff
@@ -0,0 +1,503 @@
+doesn't explicitly say `'a: 'c`. Instead, it's _inferred_ via transitivity.
+Do you see where this is going?
+
+Here's an exaggurated example of a `where` clause:
+```rust
+'a: 'b,
+'b: 'c,
+'c: 'e,
+'d: 'b,
+'e: 'd + 'f,
+```
+This looks lot like an adjacency list. That's right, this is a directed graph
+problem.
+
+If you think of lifetimes as vertices and bounds as edges pointing from longer
+to shorter lifetimes, then determining possible coercions becomes a
+simple reachability problem. The diagram below shows the exaggurated example
+as a directed graph:
+<figure>
+    <p align="center">
+        <img src="{{site.baseurl}}/images/google-internship/directed-graph.png" alt="A black-on-white image of a directed graph with 6 vertices labeled A through F. A points to B, B points to C, C points to E, E points to D and F, and D points back to B. It clearly shows that B, C, E, and D form a cycle." width="400"/>
+        <figcaption>Graph representation of lifetimes and bounds</figcaption>
+    </p>
+</figure>
+
+With this model, we can easily generalize the definition of what it means for
+some lifetime A to be able to coerce into some lifetime B. That is, lifetime A
+can coerce into lifetime B if and only if there exists a path from A to B on
+the directed graph of lifetimes and bounds. In practice, this also mean that
+all lifetimes in a cycle are all the same lifetime.
+
+So here's the gameplan:
+1. Traverse the output type, collecting any lifetimes encountered along the way.
+2. Traverse the input types. When some lifetime A is encountered, perform a
+depth-first search on shorter lifetimes while tracking visiting to break cycles.
+If one of the output lifetimes is encountered, add the parameter containing
+lifetime A to the list of parameters that the output borrows from.
+3. In code generation, use this information to communicate to the other language
+that there's a borrowing relationship as described in the previous section.
+
+Using this strategy, I was able to make Diplomat understand arbitrary lifetime
+bounds. Hey, that wasn't so bad!
+
+## Challenge 2: Borrowing from Struct Fields
+In the type philosophy section, I mentioned how non-opaque struct fields are
+publically accessible across the FFI boundary. What happens when we copy a
+borrow from a struct field?
+```rust
+#[diplomat::bridge]
+mod ffi {
+    #[diplomat::opaque]
+    pub struct Opaque {
+        // ...
+    }
+
+    pub struct Input<'a> {
+        data: &'a Opaque,
+    }
+
+    impl<'a> Input<'a> {
+        pub fn create(data: &'a Opaque) -> Self {
+            Self { data }
+        }
+
+        pub fn extract(input: Input<'a>) -> &'a Opaque {
+            let out = input.data;
+            out
+        }
+    }
+}
+```
+When rustc sees `Input::extract`, it requires that `input.data ≤ 'a ≥ out`, and
+it's able to statically enforce this because rustc is omniscient on pure Rust code.
+Diplomat, on the other hand, does FFI, where nothing is statically omniscient.
```

Comment:
suggestion: perhaps include ", since it cannot peek to the other side of the code; and most languages don't have equivalent concepts of static lifetimes anyway"